cosmos-sdk-rs/src/staking.rs

```diff
@@ -0,0 +1,240 @@
+//! Staking module support
+//!
+//! <https://docs.cosmos.network/master/modules/staking/>
+
+use crate::{
+    proto,
+    tx::{Msg, MsgType},
+    AccountId, Coin, Result,
+};
+use std::convert::{TryFrom, TryInto};
+
+/// MsgDelegate represents a message to delegate coins to a validator.
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord)]
+pub struct MsgDelegate {
+    /// Delegator's address.
+    pub delegator_address: AccountId,
+
+    /// Validator's address.
+    pub validator_address: AccountId,
+
+    /// Amount to send
+    pub amount: Option<Coin>,
```

Comment:
I'm curious... when would this ever be `None`?

Reply:
proto::cosmos::staking::v1beta1::MsgDelegate's amount can be None, so I'm mirroring that here.

Reply:
The Rust code generated from the protos has no way to express non-optional fields for messages, so that's not a great thing to go off of.

Looking at the raw Protobuf definition,  it's:

https://github.com/cosmos/cosmos-sdk/blob/9fd866e3820b3510010ae172b682d71594cd8c14/proto/cosmos/staking/v1beta1/tx.proto#L89

```proto
message MsgDelegate {
  option (gogoproto.equal)           = false;
  option (gogoproto.goproto_getters) = false;

  string                   delegator_address = 1 [(gogoproto.moretags) = "yaml:\"delegator_address\""];
  string                   validator_address = 2 [(gogoproto.moretags) = "yaml:\"validator_address\""];
  cosmos.base.v1beta1.Coin amount            = 3 [(gogoproto.nullable) = false];
}
```

It's annotated as non-nullable (using a custom gogoproto directive which the Rust proto parser doesn't understand), so this should be changed to `pub amount: Coin` and messages missing this field should be rejected.